modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C

```diff
@@ -190,9 +239,95 @@ ComputeFrictionalForceLMMechanicalContact::incorrectEdgeDroppingPost(
       continue;
 
     _weighted_gap_ptr = &_dof_to_weighted_gap[dof].first;
-    _tangential_vel_ptr = &pr.second;
+    _tangential_vel_ptr[0] = &pr.second[0];
 
-    enforceConstraintOnDof(dof);
+    if (_3d)
+    {
+      _tangential_vel_ptr[1] = &pr.second[1];
+      enforceConstraintOnDof3d(dof);
+    }
+    else
+      enforceConstraintOnDof(dof);
+  }
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::enforceConstraintOnDof3d(const DofObject * const dof)
+{
+  ComputeWeightedGapLMMechanicalContact::enforceConstraintOnDof(dof);
+
+  // Get normal LM
+  const auto normal_dof_index = dof->dof_number(_sys.number(), _var->number(), 0);
+  const ADReal & weighted_gap = *_weighted_gap_ptr;
+  ADReal contact_pressure = (*_sys.currentSolution())(normal_dof_index);
+  Moose::derivInsert(contact_pressure.derivatives(), normal_dof_index, 1.);
+
+  // Get friction LM
```

Comment:
I think we can do all of this in `enforceConstraintOnDof` and not introduce `enforceConstraintOnDof3d`. We can do:
```c++
const unsigned int num_tangents = _3d ? 2 : 1;
for (const auto i : make_range(num_tangents))
{
  // Do all the stuff, indexing into our arrays/vectors with i to get the right component
}
```
Then hopefully we can really get rid of all the `_dir` members.